src/objectives/elbo/advi.jl

```diff
@@ -0,0 +1,133 @@
+
+"""
+    ADVI(n_samples; kwargs...)
+
+Automatic differentiation variational inference (ADVI; Kucukelbir *et al.* 2017) objective.
+This computes the evidence lower-bound (ELBO) through the ADVI formulation:
+```math
+\\begin{aligned}
+\\mathrm{ADVI}\\left(\\lambda\\right)
+&\\triangleq
+\\mathbb{E}_{\\eta \\sim q_{\\lambda}}\\left[
+  \\log \\pi\\left( \\phi^{-1}\\left( \\eta \\right) \\right)
+  +
+  \\log \\lvert J_{\\phi^{-1}}\\left(\\eta\\right) \\rvert
+\\right]
++ \\mathbb{H}\\left(q_{\\lambda}\\right),
+\\end{aligned}
+```
+where ``\\phi^{-1}`` is an "inverse bijector."
+
+# Arguments
+- `n_samples::Int`: Number of Monte Carlo samples used to estimate the ELBO.
+
+# Keyword Arguments
+- `entropy`: The estimator for the entropy term. (Type `<: AbstractEntropyEstimator`; Default: ClosedFormEntropy())
+
+# Requirements
+- ``q_{\\lambda}`` implements `rand`.
+- The target `logdensity(prob, x)` must be differentiable wrt. `x` by the selected AD backend.
+
+Depending on the options, additional requirements on ``q_{\\lambda}`` may apply.
+
+# References
+* Kucukelbir, A., Tran, D., Ranganath, R., Gelman, A., & Blei, D. M. (2017). Automatic differentiation variational inference. Journal of machine learning research.
+* Titsias, M., & Lázaro-Gredilla, M. (2014, June). Doubly stochastic variational Bayes for non-conjugate inference. In International conference on machine learning (pp. 1971-1979). PMLR.
+"""
+struct ADVI{EntropyEst <: AbstractEntropyEstimator} <: AbstractVariationalObjective
+    entropy  ::EntropyEst
+    n_samples::Int
+end
+
+ADVI(n_samples::Int; entropy::AbstractEntropyEstimator = ClosedFormEntropy()) = ADVI(entropy, n_samples)
+
+Base.show(io::IO, advi::ADVI) =
+    print(io, "ADVI(entropy=$(advi.entropy), n_samples=$(advi.n_samples))")
+
+"""
+    (advi::ADVI)(
+        [rng], prob, q, zs::AbstractMatrix
+    )
```

Comment:
I thought about this, and I think it's better. I'll go with the suggestion.

Reply:
Seems like `estimate` overlaps with `Distributions`. Since it is very likely that users of this package would also have loaded `Distributions`, I'll go with `estimate_objective` to avoid annoying people.